File: ScaleformUI_Csharp/Menus/Pause Menus/Elements/Panels/MinimapPanel.cs
using ScaleformUI.PauseMenu;
using ScaleformUI.PauseMenus.Elements.Columns;
using ScaleformUI.PauseMenus.Elements.Items;

namespace ScaleformUI.PauseMenus.Elements.Panels
{
    public class MinimapPanel
    {
        internal PauseMenuBase Parent { get; set; }
        internal BaseTab ParentTab { get; set; }
        public bool HidePedBlip { get; set; } = true;
        internal Vector2 mapPosition = new Vector2();
        internal float zoomDistance = 0;
        internal bool enabled;
        private bool turnedOn = false;
        private bool IsRadarVisible = !IsRadarHidden();
        internal int localCoronaMapStage = -1;

        public MinimapRoute MinimapRoute;
        public List<FakeBlip> MinimapBlips { get; internal set; }

        public MinimapPanel(BaseTab parenttab)
        {
            MinimapBlips = new List<FakeBlip>();
            MinimapRoute = new MinimapRoute();
            ParentTab = parenttab;
        }

        public bool Enabled
        {
            get => enabled;
            set
            {
                if (Parent != null && Parent.Visible)
                {
                    if(ParentTab is PlayerListTab plTab)
                    {
                        if(plTab.CurrentColumn is PlayerListColumn col)
                        {
                            if (col.CurrentItem.KeepPanelVisible)
                            {
                                return;
                            }
                        }
                    }
                }
                enabled = value;
                if (enabled)
                {
                    if (localCoronaMapStage == -1)
                        localCoronaMapStage = 0;
                }
                else
                {
                    localCoronaMapStage = -1;
                    if (turnedOn)
                    {
                        IsRadarVisible = !IsRadarHidden();
                        DisplayRadar(false);
                        RaceGalleryFullscreen(false);
                        turnedOn = false;
                    }
                }
                if (Parent != null && Parent.Visible && ParentTab.Visible)
                {
                    if (ParentTab is PlayerListTab plTab)
                    {
                        if (plTab.CurrentColumn is PlayerListColumn col)
                        {
                            if (value)
                            {
                                col.CurrentItem.Dispose();
                            }
                            else
                            {
                                col.CurrentItem.CreateClonedPed();
                            }
                        }
                        var rightCol = ParentTab.RightColumn;
                        if (rightCol is MissionDetailsPanel panel)
                            panel.ColumnVisible = !enabled && plTab.CurrentColumn is not PlayerListColumn;
                    }
                }
            }
        }

         internal void InitializeMapSize()
        {
            float top = float.NegativeInfinity;
            float bottom = float.PositiveInfinity;
            float left = float.PositiveInfinity;
            float right = float.NegativeInfinity;

            foreach (var data in MinimapRoute.CheckPoints)
            {
                top = Math.Max(top, data.Position.Y);
                bottom = Math.Min(bottom, data.Position.Y);
                left = Math.Min(left, data.Position.X);
                right = Math.Max(right, data.Position.X);
            }

            top = Math.Max(top, MinimapRoute.StartPoint.Position.Y);
            bottom = Math.Min(bottom, MinimapRoute.StartPoint.Position.Y);
            left = Math.Min(left, MinimapRoute.StartPoint.Position.X);
            right = Math.Max(right, MinimapRoute.StartPoint.Position.X);

            top = Math.Max(top, MinimapRoute.EndPoint.Position.Y);
            bottom = Math.Min(bottom, MinimapRoute.EndPoint.Position.Y);
            left = Math.Min(left, MinimapRoute.EndPoint.Position.X);
            right = Math.Max(right, MinimapRoute.EndPoint.Position.X);

            Vector3 topLeft = new Vector3(left, top, 0);
            Vector3 bottomRight = new Vector3(right, bottom, 0);

            // Center of square area
            mapPosition = new Vector2((topLeft.X + bottomRight.X) / 2, (topLeft.Y + bottomRight.Y) / 2);

            // Calculate our range and get the correct zoom.
            float DistanceX = Math.Abs(left - right);
            float DistanceY = Math.Abs(top - bottom);

            if (DistanceX > DistanceY)
            {
                zoomDistance = DistanceX / 1.5f;
            }
            else
            {
                zoomDistance = DistanceY / 2.0f;
            }

            RefreshMapPosition(mapPosition);
            LockMinimapAngle(0);

            //!! Draw Debug
            // var blipArea = AddBlipForArea(mapPosition.X, mapPosition.Y, 0.0f, DistanceX, DistanceY);
            // SetBlipAlpha(blipArea, 150);
            // RaceGalleryNextBlipSprite(1);
            // var blipTop = RaceGalleryAddBlip(topLeft.X, topLeft.Y, 0.0f);
            // RaceGalleryNextBlipSprite(1);
            // var blipBottom = RaceGalleryAddBlip(bottomRight.X, bottomRight.Y, 0.0f);
            // ShowNumberOnBlip(blipTop, 1);
            // ShowNumberOnBlip(blipBottom, 2);
        }

        public void RefreshMapPosition(Vector2 position)
        {
            mapPosition = new Vector2(position);
            if (ParentTab is GalleryTab g)
                zoomDistance = g.bigPic ? 600 : 1200;
        }

        internal void SetupBlips()
        {
            foreach (FakeBlip blip in MinimapBlips)
            {
                RaceGalleryNextBlipSprite(blip.Sprite);
                int b = RaceGalleryAddBlip(blip.Position.X, blip.Position.Y, blip.Position.Z);
                if (blip.Scale > 0)
                    SetBlipScale(b, blip.Scale);
                SetBlipColour(b, (int)blip.Color);
            }
        }

        public void MaintainMap()
        {
            switch (localCoronaMapStage)
            {
                case 0:
                    InitializeMap();
                    break;
                case 1:
                    ProcessMap();
                    break;
            }
        }
        public async Task ProcessMap()
        {
            if (enabled)
            {
                if (!turnedOn)
                {
                    DisplayRadar(IsRadarVisible);
                    RaceGalleryFullscreen(true);
                    turnedOn = true;
                }
            }
            else
            {
                if (turnedOn)
                {
                    IsRadarVisible = !IsRadarHidden();
                    DisplayRadar(false);
                    RaceGalleryFullscreen(false);
                    turnedOn = false;
                    Dispose();
                }
            }
            if(HidePedBlip)
                SetPlayerBlipPositionThisFrame(-5000, -5000);
            RefreshZoom();
        }

        internal void InitializeMapDisplay()
        {
            DeleteWaypoint();
            SetWaypointOff();
            ClearGpsCustomRoute();
            ClearGpsMultiRoute();
            SetPoliceRadarBlips(false);

            MinimapRoute.SetupCustomRoute();
            SetupBlips();
        }

        internal void InitializeMap()
        {
            // Use the data to set up our map.
            InitializeMapSize();
            // Initialise all our blips from data
            InitializeMapDisplay();
            // Set the zoom now
            RefreshZoom();
            localCoronaMapStage = 1;
        }

        /// <summary>
        /// To be called every tick
        /// </summary>
        internal void RefreshZoom()
        {
            if (zoomDistance != 0)
                SetRadarZoomToDistance(zoomDistance);
            LockMinimapPosition(mapPosition.X, mapPosition.Y);
        }

        internal void Dispose()
        {
            localCoronaMapStage = 0;
            enabled = false;
            N_0x2de6c5e2e996f178(0);
            SetPoliceRadarBlips(true);
            DisplayRadar(IsRadarVisible);
            RaceGalleryFullscreen(false);
            ClearRaceGalleryBlips();
            zoomDistance = 0;
            SetRadarZoom(0);
            SetGpsCustomRouteRender(false, 18, 30);
            SetGpsMultiRouteRender(false);
            UnlockMinimapPosition();
            UnlockMinimapAngle();
            DeleteWaypoint();
            ClearGpsCustomRoute();
            ClearGpsFlags();
            MinimapBlips.Clear();
            MinimapRoute = new MinimapRoute();
            SetBigmapActive(false, false);
        }

        public void ClearMinimap()
        {
            MinimapBlips.Clear();
            MinimapRoute = new MinimapRoute();
            localCoronaMapStage = 0;
            zoomDistance = 0;
            ClearRaceGalleryBlips();
            SetRadarZoom(0);
            SetGpsCustomRouteRender(false, 18, 30);
            DeleteWaypoint();
            ClearGpsCustomRoute();
            ClearGpsFlags();
            SetBigmapActive(false, false);
        }
    }
}

File: ScaleformUI_Lua/src/Menus/PauseMenus/Elements/Panels/MinimapPanel.lua
MinimapPanel = setmetatable({}, MinimapPanel)
MinimapPanel.__index = MinimapPanel
MinimapPanel.__call = function()
    return "Panel", "MinimapPanel"
end

---@class MinimapPanel
---@field private enabled boolean
---@field private turnedOn boolean
---@field private mapPosition vector2
---@field private localMapStage number
---@field private IsRadarVisible boolean
---@field private New fun(parent:BaseTab)
---@field public Parent BaseTab
---@field public MinimapBlips FakeBlip[]
---@field public MinimapRoute MinimapRoute
---@field public Enabled fun(_e: boolean|nil)
---@field private InitializeMapSize fun()
---@field private SetupBlips fun()
---@field private MaintainMap fun()
---@field private ProcessMap fun()
---@field private InitializeMapDisplay fun()
---@field private InitializeMap fun()
---@field private RefreshZoom fun()
---@field private Dispose fun()
---@field public ClearMinimap fun()

function MinimapPanel.New(parentTab)
    local _data = {
        Parent = nil,
        ParentTab = parentTab,
        MinimapBlips = {},
        MinimapRoute = MinimapRoute.New(),
        mapPosition = vector2(0, 0),
        enabled = false,
        turnedOn = false,
        localMapStage = -1,
        HidePedBlip = false,
        IsRadarVisible = not IsRadarHidden(),
    }
    return setmetatable(_data, MinimapPanel)
end

function MinimapPanel:Enabled(_e)
    if _e == nil then
        return self.enabled
    else
        if self.Parent ~= nil and self.Parent:Visible() then
            if self.ParentTab() == "PlayerListTab" and self.ParentTab:CurrentColumn().type == PLT_COLUMNS.PLAYERS then
                if self.ParentTab:CurrentColumn():CurrentItem():KeepPanelVisible() then
                    self:Dispose()
                    return
                end
            end
        end
        self.enabled = ToBool(_e)
        if _e == true then
            if self.localMapStage == -1 then
                self.localMapStage = 0
            end
        else
            self.localMapStage = -1
            if self.turnedOn then
                self.IsRadarVisible = IsMinimapRendering()
                DisplayRadar(false)
                RaceGalleryFullscreen(false)
                self.turnedOn = false
            end
        end
        if self.Parent ~= nil and self.Parent:Visible() then
            if self.ParentTab() == "PlayerListTab" then
                if self.ParentTab:CurrentColumn().type == PLT_COLUMNS.PLAYERS then
                    if ToBool(_e) then
                        self.ParentTab:CurrentColumn():CurrentItem():Dispose()
                    else
                        self.ParentTab:CurrentColumn():CurrentItem():AddPedToPauseMenu()
                    end
                end
                if self.ParentTab.RightColumn ~= nil and self.ParentTab.RightColumn.type == PLT_COLUMNS.MISSION_DETAILS then
                    self.ParentTab.RightColumn:ColumnVisible(not ToBool(_e) and
                    self.ParentTab:CurrentColumn().type ~= PLT_COLUMNS.PLAYERS)
                end
            end
        end
    end
end

function MinimapPanel:InitializeMapSize()
    local top = -math.huge
    local bottom = math.huge
    local left = math.huge
    local right = -math.huge

    for k, data in pairs(self.MinimapRoute.CheckPoints) do
        top = math.max(top, data.Position.y)
        bottom = math.min(bottom, data.Position.y)
        left = math.min(left, data.Position.x)
        right = math.max(right, data.Position.x)
    end

    top = math.max(top, self.MinimapRoute.StartPoint.Position.y)
    bottom = math.min(bottom, self.MinimapRoute.StartPoint.Position.y)
    left = math.min(left, self.MinimapRoute.StartPoint.Position.x)
    right = math.max(right, self.MinimapRoute.StartPoint.Position.x)

    top = math.max(top, self.MinimapRoute.EndPoint.Position.y)
    bottom = math.min(bottom, self.MinimapRoute.EndPoint.Position.y)
    left = math.min(left, self.MinimapRoute.EndPoint.Position.x)
    right = math.max(right, self.MinimapRoute.EndPoint.Position.x)

    local topLeft = vector3(left, top, 0)
    local bottomRight = vector3(right, bottom, 0)

    -- Center of square area
    self.mapPosition = vector2((topLeft.x + bottomRight.x) / 2, (topLeft.y + bottomRight.y) / 2)

    -- Calculate our range and get the correct zoom.
    local DistanceX = math.abs(left - right)
    local DistanceY = math.abs(top - bottom)
    local Diagonal = math.sqrt(DistanceX ^ 2 + DistanceY ^ 2)

    if DistanceX == DistanceY then      -- Square
        self.zoomDistance = Diagonal / 2.4
    elseif (DistanceX > DistanceY) then -- Horizontal
        local mul = 1.7
        self.zoomDistance = math.min(DistanceX / mul, Diagonal / mul)
    else -- Vertical
        local mul = 2.4
        self.zoomDistance = math.max(DistanceY / mul, Diagonal / mul)
    end

    self:RefreshMapPosition(self.mapPosition)
    LockMinimapAngle(0)

    --!! Draw Debug
    -- local blipArea = AddBlipForArea(self.mapPosition.x, self.mapPosition.y, 0.0, DistanceX, DistanceY)
    -- SetBlipAlpha(blipArea, 150)
    -- RaceGalleryNextBlipSprite(1)
    -- local blipTop = RaceGalleryAddBlip(topLeft.x, topLeft.y, 0.0)
    -- RaceGalleryNextBlipSprite(1)
    -- local blipBottom = RaceGalleryAddBlip(bottomRight.x, bottomRight.y, 0.0)
    -- ShowNumberOnBlip(blipTop, 1)
    -- ShowNumberOnBlip(blipBottom, 2)
end

function MinimapPanel:RefreshMapPosition(position)
    if position == vec(0,0) then return end
    self.mapPosition = position
    LockMinimapPosition(self.mapPosition.x, self.mapPosition.y)
    if self.ParentTab() == "GalleryTab" then
        if self.ParentTab.bigPic then
            self.zoomDistance = 600.0
        else
            self.zoomDistance = 1200.0
        end
    end
end

function MinimapPanel:SetupBlips()
    for _, blip in pairs(self.MinimapBlips) do
        RaceGalleryNextBlipSprite(blip.Sprite)
        local b = RaceGalleryAddBlip(blip.Position.x, blip.Position.y, blip.Position.z)
        if blip.Scale > 0 then
            SetBlipScale(b, blip.Scale) 
        end
        SetBlipColour(b, blip.Color)
    end
end

function MinimapPanel:MaintainMap()
    if self.localMapStage == 0 then
        self:InitializeMap()
    elseif self.localMapStage == 1 then
        self:ProcessMap()
    end
end

function MinimapPanel:ProcessMap()
    if self.enabled then
        if not self.turnedOn then
            DisplayRadar(self.IsRadarVisible)
            RaceGalleryFullscreen(true)
            self.turnedOn = true
        end
    else
        if self.turnedOn then
            self.IsRadarVisible = not IsRadarHidden()
            DisplayRadar(false)
            RaceGalleryFullscreen(false)
            self.turnedOn = false
            self:Dispose()
        end
    end
    if self.HidePedBlip then
        SetPlayerBlipPositionThisFrame(-5000.0, -5000.0)
    end
    self:RefreshZoom()
end

function MinimapPanel:InitializeMapDisplay()
    DeleteWaypoint()
    SetWaypointOff()
    ClearGpsCustomRoute()
    ClearGpsMultiRoute()
    SetPoliceRadarBlips(false)

    self.MinimapRoute:SetupCustomRoute()
    self:SetupBlips()
end

function MinimapPanel:InitializeMap()
    -- Use the data to set up our map.
    self:InitializeMapSize()
    -- Initialise all our blips from data
    self:InitializeMapDisplay()
    -- Set the zoom now
    self:RefreshZoom()
    self.localMapStage = 1
end

function MinimapPanel:RefreshZoom()
    if self.zoomDistance ~= 0 then
        SetRadarZoomToDistance(self.zoomDistance)
    end
    LockMinimapPosition(self.mapPosition.x, self.mapPosition.y)
end

function MinimapPanel:Dispose()
    self.localMapStage = 0
    self.enabled = false
    SetPoliceRadarBlips(true)
    PauseToggleFullscreenMap(false)
    DisplayRadar(self.IsRadarVisible)
    RaceGalleryFullscreen(false)
    ClearRaceGalleryBlips()
    self.zoomDistance = 0
    SetRadarZoom(0)
    SetGpsCustomRouteRender(false, 18, 30)
    SetGpsMultiRouteRender(false)
    UnlockMinimapPosition()
    UnlockMinimapAngle()
    DeleteWaypoint()
    ClearGpsCustomRoute()
    ClearGpsFlags()
    SetBigmapActive(false, false)
    self.MinimapBlips = {}
    self.MinimapRoute = MinimapRoute.New()
end

function MinimapPanel:ClearMinimap()
    self.MinimapBlips = {}
    self.MinimapRoute = MinimapRoute.New()
    self.localMapStage = 0
    self.zoomDistance = 0
    ClearRaceGalleryBlips()
    SetRadarZoom(0)
    SetGpsCustomRouteRender(false, 18, 30)
    DeleteWaypoint()
    ClearGpsCustomRoute()
    ClearGpsFlags()
    SetBigmapActive(false, false)
end
